bindings/py/nupic/bindings/regions/PyRegion.py

```diff
@@ -0,0 +1,359 @@
+  * setParameter
+  * setParameterArray
+  * getOutputElementCount
+
+  The getSpec is a class method, which is actually required but since it's
+  not an instance method the @abstractmethod decorator doesn't apply.
+
+  Finally, PyRegion provides reasonable default implementation to some methods.
+  Sub-classes may opt to override these methods or use the default
+  implementation (often recommended).
+
+  The implemented methods are:
+
+  * getParameter
+  * getParameterArray
+  * getParameterArrayCount
+  * executeMethod
+
+  """
+  __metaclass__ = ABCMeta
+
+
+  @classmethod
+  def getSpec(cls):
+    """Returns the region spec for this region. The Region Spec is a dictionary
+    with the following keys:
+    description -- a string
+
+    singleNodeOnly -- a boolean (True if this Region supports only a single node)
+
+    inputs -- a dictionary in which the keys are the names of the inputs and
+    the values are dictionaries with these keys:
+         description - string
+         regionLevel -- True if this is a "region-level" input.
+         dataType - a string describing the data type, usually 'Real32'
+         count - the number of items in the input. 0 means unspecified.
+         required -- boolean - whether the input is must be connected
+         isDefaultInput -- must be True for exactly one input
+         requireSplitterMap -- [just set this to False.]
+
+    outputs -- a dictionary with similar structure to inputs. The keys
+    are:
+         description
+         dataType
+         count
+         regionLevel
+         isDefaultOutput
+
+    parameters -- a dictionary of dictionaries with the following keys:
+         description
+         dataType
+         count
+         constraints (optional)
+         accessMode (one of "ReadWrite", "Read", "Create")
+
+    This class method is called by NuPIC before creating a Region.
+    """
+    raise NotImplementedError()
+
+
+  @abstractmethod
+  def __init__(self, *args, **kwars):
+    """Initialize the node with creation parameters from the node spec
+
+    Should be implemented by subclasses (unless there are no creation params)
+    """
+
+
+  @abstractmethod
+  def initialize(self, inputs, outputs):
+    """Initialize the node after the network is fully linked
+
+    It is called once by NuPIC before the first call to compute(). It is
+    a good place to perform one time initialization that depend on the inputs
+    and/or outputs. The region may also remember its inputs and outputs here
+    because they will not change.
+
+    inputs: dict of numpy arrays (one per input)
+    outputs: dict of numpy arrays (one per output)
+    """
+
+
+  @abstractmethod
+  def compute(self, inputs, outputs):
+    """Perform the main computation
+
+    This method is called in each iteration for each phase the node supports.
+
+    inputs: dict of numpy arrays (one per input)
+    outputs: dict of numpy arrays (one per output)
+    """
+
+
+  def getOutputElementCount(self, name):
```

Comment:
If you read the class description it states that not all methods have to be implemented. So it is more of an optional method than an abstract method. Maybe not good practice but I'm just copy/pasting this code (for the most part).